### PR TITLE
Allow KillPod to take a gracePeriodOverride

### PR DESCRIPTION
**Pull request by derekwaynecarr**
The proposal for pod eviction (see https://github.com/kubernetes/kubernetes/pull/18724) will need to allow a pod to be killed immediately if a hard eviction threshold has been met.  In that case, we will want to call `KillPod` with a `gracePeriodOverride=0` so we can do `SIGKILL` and reclaim resources quickly.  In the case of a soft eviction threshold having been met, we discussed letting an operator set a maximum soft eviction pod termination grace period, in that case we would call with a `gracePeriodOverride` set to that value.

This PR adds the new argument for `gracePeriodOverride` and plumbs its support in the code for `docker`, and a TODO for `rkt`.

/cc @vishh  - per our slack chat

/fyi @kubernetes/sig-node @kubernetes/rh-cluster-infra

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
@vishh - poke ;-)

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -1354,6 +1354,10 @@ func (dm *DockerManager) killContainer(containerID kubecontainer.ContainerID, co
 	if gracePeriod < minimumGracePeriodInSeconds {
 		gracePeriod = minimumGracePeriodInSeconds
 	}
+	if gracePeriodOverride != nil {
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Add a comment here about why gracePeriodOverride is allowed to go below minimum (because this is a hard kill) and ONLY hard kill paths are allowed to do this.  I.e. no one else in the entire Kubelet should ever set this value less than minimumGracePeriodInSeconds, otherwise they can and will corrupt user data.

Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Although now that I think about it - is minimum really our problem here?

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -1304,12 +1304,12 @@ func (dm *DockerManager) KillContainerInPod(containerID kubecontainer.ContainerI
...
 		}
 	}
-	return dm.killContainer(containerID, container, pod, message)
+	return dm.killContainer(containerID, container, pod, message, gracePeriodOverride)
 }
 
 // killContainer accepts a containerID and an optional container or pod containing shutdown policies. Invoke
 // KillContainerInPod if information must be retrieved first.
-func (dm *DockerManager) killContainer(containerID kubecontainer.ContainerID, container *api.Container, pod *api.Pod, reason string) error {
+func (dm *DockerManager) killContainer(containerID kubecontainer.ContainerID, container *api.Container, pod *api.Pod, reason string, gracePeriodOverride *int64) error {
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Godoc on why no one should use gracePeriodOverride

File: pkg/kubelet/dockertools/manager.go (hunk 1 above)
Comment (derekwaynecarr) [inline pkg/kubelet/dockertools/manager.go]: @smarterclayton - 2s is too long when running out of resources.  Are you asking why do we have a minimum at all in the normal case?

Comment (derekwaynecarr) [inline pkg/kubelet/dockertools/manager.go]: btw, systemd defaults to 90s for default stop timeout, not sure why we settled on 2s here...

https://www.freedesktop.org/software/systemd/man/systemd-system.conf.html

Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Defaults are handled by Kube API.  This is the fundamental protection that
we always give EVERYONE at least 2s to respond to sigterm.  Even if you say
0s.

I'm fine for this to bypass grace period and kill directly.  I just wanted
to have the comment recorded.

File: pkg/kubelet/container/runtime.go [hunk 3]
```diff
@@ -79,7 +79,9 @@ type Runtime interface {
 	SyncPod(pod *api.Pod, apiPodStatus api.PodStatus, podStatus *PodStatus, pullSecrets []api.Secret, backOff *flowcontrol.Backoff) PodSyncResult
 	// KillPod kills all the containers of a pod. Pod may be nil, running pod must not be.
 	// TODO(random-liu): Return PodSyncResult in KillPod.
-	KillPod(pod *api.Pod, runningPod Pod) error
+	// gracePeriodOverride if specified allows the caller to override the pod default grace period.
+	// it is usueful when doing SIGKILL for hard eviction scenarios, or max grace period during soft eviction scenarios.
+	KillPod(pod *api.Pod, runningPod Pod, gracePeriodOverride *int64) error
```
Comment (vishh) [inline pkg/kubelet/container/runtime.go]: nit: Why not a `*time.Duration` instead of `*int64`?

File: pkg/kubelet/rkt/rkt.go [hunk 4]
```diff
@@ -1413,7 +1414,9 @@ func (r *Runtime) SyncPod(pod *api.Pod, podStatus api.PodStatus, internalPodStat
 	if restartPod {
 		// Kill the pod only if the pod is actually running.
 		if len(runningPod.Containers) > 0 {
-			if err = r.KillPod(pod, runningPod); err != nil {
+			// we stick with the default gracePeriod
```
Comment (vishh) [inline pkg/kubelet/rkt/rkt.go]: cc @yifan-gu

**Comment by vishh** [comment]
This PR LGTM

**Comment by derekwaynecarr** [comment]
@smarterclayton - added the requested godoc.  this is blocking eviction work, can I get a lgtm?

**Comment by vishh** [comment]
@derekwaynecarr If @smarterclayton doesn't have any more comments, I'm ok with merging this PR

File: pkg/kubelet/container/runtime.go (hunk 3 above)
Comment (vishh) [inline pkg/kubelet/container/runtime.go]: ping @derekwaynecarr ^^

**Comment by smarterclayton** [comment]
I'm ok with it

**Comment by derekwaynecarr** [comment]
Since everyone claimed to be ok with merging this, I am applying the tag.

**Comment by vishh** [comment]
@derekwaynecarr I think you missed https://github.com/kubernetes/kubernetes/pull/24843#discussion_r61780865.

File: pkg/kubelet/container/runtime.go (hunk 3 above)
Comment (derekwaynecarr) [inline pkg/kubelet/container/runtime.go]: I went with int for the following reasons:

*  it was consistent with pod termination grace period seconds (int64)
 * https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1208
* it was consistent with docker client StopContainer timeout in seconds (int)
 * https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/docker.go#L64
* it was consistent with systemd `TimeoutStopSec`
 * https://www.freedesktop.org/software/systemd/man/systemd.service.html

so basically, taking a duration means we have to remember to convert to second precision which seemed more complicated/error prone

**Comment by derekwaynecarr** [comment]
removing the LGTM because @vishh desired grace period to use time.Duration, will poke when updated.

**Comment by smarterclayton** [comment]
All grace periods are ints, everywhere in the system.  Would prefer not to
do duration, and instead keep it seconds.

**Comment by derekwaynecarr** [comment]
@smarterclayton - OMG, I just pushed the change now ;-)

**Comment by derekwaynecarr** [comment]
I reverted back to *int64 and am re-applying LGTM.
